src/lib/cpu-clk-manager.c

```diff
@@ -0,0 +1,95 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2022 Intel Corporation. All rights reserved.
+//
+// Author: Krzysztof Frydryk <frydryk.krzysztof@intel.com>
+
+#include <rtos/spinlock.h>
+#include <sof/sof.h>
+#include <stdint.h>
+#include <sof/lib/cpu-clk-manager.h>
+#include <rtos/clk.h>
+#include <sof/sof.h>
+#include <sof/math/numbers.h>
+#include <errno.h>
+#ifdef __ZEPHYR__
+#include <zephyr/sys/util.h>
+#endif /* __ZEPHYR__ */
+
+static struct kcps_budget_data kcps_data;
+
+static int request_freq_change(int core, int freq)
+{
+	int current_freq;
+	int selected_freq_id;
+	struct clock_info *clk;
+
+	clk = clocks_get() + core;
+
+	for (selected_freq_id = 0; selected_freq_id < clk->freqs_num; selected_freq_id++) {
+		if (freq < clk->freqs[selected_freq_id].freq)
+			break;
+	}
+
+	/* don't change clock frequency if already using proper clock */
+	current_freq = clock_get_freq(core);
+	if (clk->freqs[selected_freq_id].freq != current_freq)
+		clock_set_freq(core, freq);
+
+	return 0;
+}
+
+static int max_core_consumption(void)
+{
+	int result = 0;
+	int core;
+
+	for (core = 0; core < CONFIG_CORE_COUNT; core++)
+		result = MAX(result, kcps_data.kcps_consumption[core]);
+
+	return result;
+}
+
+int core_kcps_adjust(int core, int kcps_delta)
```

Comment:
Not sure why `core` is of signed int type (also in many other places in the code) while this piece of the code (as many others) works only for sequential values in 0..CONFIG_CORE_COUNT-1 range. Note that `cpu_get_id()` is now mapped to Zephyr's `arch_proc_id()` that returns `uint32_t`. Note that the latter does not guarantee the condition mentioned above (see https://docs.zephyrproject.org/latest/hardware/porting/arch.html#c.arch_proc_id), so indexing `kcps_consumption` with `core` might not be safe, unless there is some integration note but I did not find anything in place the `cpu_get_id()` is declared/defined.

Reply:
That's the way it's done in other parts of code. I also couldn't find any spec guaranteeing core ids to be continues and starting at 0. I can imagine something like platform specific iterator for enumerating cores, but that is probably not the place for such discussion. And it would require refactoring other code too.
I did change the iterators to be unsigned, just for the sake of it :)